### DISCUSSION TITLE
Preparation for 2.0

### DISCUSSION THREAD

**Discussion by olivermrbl** [medusajs/medusa#7195] [Announcements]
Things are progressing well toward Medusa 2.0, and we are excited to share an early preview version for experimentation and exploration purposes next week. 

**The preview version will not be production-ready. However, you can start developing with it to get acquainted with Medusa 2.0.**

Medusa 2.0 introduces architectural changes to our product, primarily driven by fully decoupling our modules to eliminate interdependencies between them. This new architecture will enable greater composability, use of standalone modules, and support for incremental adoption. Refer to our recent [Recap](https://medusajs.com/recap) for more details about the architectural changes and other features in 2.0.

Our new architecture has led to several changes to our open-source repository, including:
- Reorganization of packages
- Deprecation of plugins
- Introduction of an LTS branch

This document is here to help you understand the implications of these changes, which have taken effect today.

### **Reorganization of packages**

We will use Medusa 2.0 as an opportunity to improve the discoverability and navigation of our repository. This is causing structural changes to our folder structure.

Going forward, the repository will be structured as follows:

```tsx
/packages
  /modules
    /cart
    /product
    ...

  /admin
    /dashboard

  /core
    /workflows-sdk
    /modules-sdk
    /medusa-js
    ...
```

Many packages have been left out for the sake of brevity. The purpose is to illustrate the new subfolders in `/packages`. 

This project restructuring is subject to change as we move closer to the official release, but you should consider this a good proxy for what it will look like.

### Deprecation of plugins

We have reviewed our suite of officially supported plugins and decided to deprecate several. Some are broken. Some will not be supported in 2.0. And some are deprecated to eliminate a maintenance burden for our core team. 

If you wish to take over the maintenance of any of the below-mentioned plugins, please leave a comment, and we can discuss. 

The following plugins will be deprecated and removed from our repository in 2.0:
- `medusa-source-shopify`
- `gatsby-source-medusa`
- `medusa-payment-adyen`
- `medusa-plugin-economic`
- `medusa-plugin-mailchimp`
- `medusa-plugin-twilio-sms`

The following plugins will be migrated to a different package in 2.0:
- `medusa-file-local` → `@medusajs/file-local`
- `medusa-file-s3` → `@medusajs/file-s3`
- `medusa-file-spaces` → `@medusajs/file-s3`
- `medusa-file-minio` → `@medusajs/file-s3`
- `medusa-fulfillment-manual` → `@medusajs/fulfillment-manual`
- `medusa-fulfillment-webshipper` → `@medusajs/fulfillment-webshipper`
- `medusa-payment-klarna` → `@medusajs/payment-klarna`
- `medusa-payment-manual` → `@medusajs/payment-manual`
- `medusa-payment-stripe` → `@medusajs/payment-stripe`
- `medusa-payment-paypal` → `@medusajs/payment-paypal`

Migrating from the existing plugins to the new should be more or less frictionless, as the functionality stays the same. Only the configuration in `medusa-config.js` will change; we will share this later.

Consider both lists non-exhaustive, as these are also subject to change before the official release.

### Introduction of v1.x branch

From today, we will no longer actively update Medusa v1.x with new features and/or larger improvements. Our core team will only patch security vulnerabilities and critical issues. External contributors are still more than welcome to introduce improvements and fix bugs. However, such PRs will be considered low priority for our team to review and will take longer to land in an official release than previously.

All changes to versions 1.x should use a new maintenance branch: `v.1x`. 

Going forward, there will be two release tracks:

- `develop` → preview releases of Medusa 2.0
- `v1.x` → official releases of Medusa 1.x

If you wish to resolve an issue in v1.x, please ensure your changes and pull request use `v1.x` as the base branch. 

**A comment on currently open PRs**

If you have an open PR to `develop`, we kindly ask you to rebase off of `v1.x` and change the base branch on the pull request. We will attend to it as soon as possible. 

Please don’t hesitate to ask questions or voice concerns about these new initiatives. We will address them to the best of our ability.

**Comment by minhchu** [comment]
Hi @olivermrbl 

Thanks for your amazing work.

Could you take a look at this issue: https://github.com/medusajs/medusa/issues/7156 
It is blocking us to production

**Comment by genox** [comment]
Hi @olivermrbl 

I wonder if medusa v2 will bring any new features to the admin, in terms of editing product information (like localised description and titles, multiple image variants, etc). I am currently facing the decision whether I should focus on moving catalog management entirely to a headless cms insteaf of extending core functionality of medusa, since it is kinda hard to extend the existing views in the admin and I would rather not build complicated product entities in medusa v1 as this has proven more complex in the past than I was hoping.

**Reply by olivermrbl** [reply]
Could I get you to share the specific features you would like to see added to the products domain? Some might be on the roadmap. 

Localised titles and descriptions will not be part of the first iteration of the product module. Variants images are under consideration.

What about extending existing views in admin do you find hard? Would love to know how we can improve.

**Reply by muhammadx2k** [reply]
Hi @olivermrbl, the only thing I find hard to extend anything on current version is that the documentation is not very good. It's very ambiguous and confusing, also looks very hard. 

I would like to know how to extend the modules to have a subscription based business running on Medusa. 

So, if you could provide a improved documentation with v2, it will be awesome! 

Thanks for the amazing work.

**Reply by danvayn** [reply]
You can find that information (on pre-2.0 Medusa) here:
[https://docs.medusajs.com/recipes/subscriptions](https://docs.medusajs.com/recipes/subscriptions)

I actually think the documentation is pretty good -- Ecommerce is just complex.

I would love to see more e2e tutorials though. Youtube content would be awesome.

**Reply by genox** [reply]
@olivermrbl If I were to make a wishlist for medusa, it would look something like that:

- "Schema in code" like strapi, payload, sanity and others. No need to manually add api endpoints, types, validity checks, UI. data import and export.
- A more complete Admin experience. Right now the common opinion seems to be to rely on a 3rd party CMS for almost any use case that extends over what Medusa offers out of the box. Or at least, this is the idea that is floating around. Basically, look at the existing ecommerce backends and offer the functionality from the beginning or provide solutions for all major headless cms out of the box.
- Better logging, better diagnostics. No matter what I do, there are places in the stack where logging does not work and I have no idea why. The system feels a bit intransparent as it is now. E.g. sometimes subscribers are not picked up because of a seemingly arbitrary naming requirement? Stuff like that. It costs a lot of time and effort to get past issues like this and they are really unnecessary.
- Less "expensive" stack options: I would like to see the inclusion of SQlite and redis alternatives that make it possible for the entire stack to run more efficiently for small operations. No idea how that looks like but something I was wondering lately after looking at an endless list of services for multiple staging environments. :)

**Reply by SonnyFishback** [reply]
@olivermrbl, I would like to see decoupling of authentication/session management from Medusa to make it easier to use managed identity providers without having to implement hacky overrides  in Medusa. For example, I want to use Supabase for my DB however, it is not worth it without using Supabase DBB without Supabase Auth and Supababse DB Row Level Security.

Conversation #5251 
PR #5262

**Reply by olivermrbl** [reply]
> So, if you could provide a improved documentation with v2, it will be awesome!

@rk12129, we are working on something big for the 2.0 documentation. Hope to share more very soon. I noted down your request and we will consider it for an example implementation.

**Reply by olivermrbl** [reply]
@genox, thank you very much for your input – really appreciate it. I'll try to comment on a few of the items in-line here:

> "Schema in code"

This is on the drawing board. Still early days, so I can't share anything yet, but we are exploring this as we speak. If there are any specific tools you would like us to consider in this context, please let me know, as now would be a very good time to have that.

> A more complete Admin experience

What specifically are you looking for? Our UI Routes and Widgets allow you to extend the admin dashboard with entire domains, e.g. blog posts. You should be able to build the features you would otherwise need a CMS for – naturally not as feature-rich, but it should take you pretty far.

> Better logging, better diagnostics

Will be improved in 2.0!

**Reply by bert-rstk** [reply]
@olivermrbl for a schema-in-code CMS I'd suggest looking closely at PayloadCMS.

It aligns with MedusaJS "philosophically" in many aspects: TypeScript first, headless and modular, customizable Admin and UI components, lightweight footprint for serverless deployments to Vercel, closely integrates with NextJS and RSC (see recent 3.0 release). A noteworthy difference is that it was built on MongoDB, with PostgreSQL only added recently. Drizzle is the ORM for PostgreSQL, which differs from Medusa's move from TypeORM to MikroORM.

I agree with @genox that CMS-like functionality is key to easily create and extend custom attributes/entities in Medusa. IMHO this not only helps compete with Shopify on features (custom fields), but supports Medusa's broader mission to offer the most easily customizable "full-stack" e-commerce platform. The upcoming workflow engine/UI is a great example of "lowering the barrier" to customization 💪

**Reply by HarleySalas** [reply]
I totally agree here. I've been using Payload a lot lately and I'm planning to make a 2-way sync plugin, though it'd be nice to see a more community driven, or officially supported approach to this. I totally agree that it's best to have content management systems do what they do best; content management and medusa what it does best; e-commerce. I want to see both Medusa and Payload grow and as it stands, while Medusa and Payload are both very powerful, it's still quite intimidating, especially for less experienced developers to tackle these sort of things. I'd love to see what beautiful thing could be born when the experts of these platforms either build a plugin, or perhaps even just make some great documentation that shows the best approaches on how to do it.

**Reply by bert-rstk** [reply]
Integrating Payload -or a preferred CMS- as an official plugin in Medusa would be great. I believe many Medusa users connect it to an external CMS to build custom entities for their content (CMS) and product (PIM) needs. Payload could be built as a module via its [Local API](https://payloadcms.com/docs/local-api/overview), for a closer and latency-free integration.

@genox raises a more general point about the DX of a **CMS and app framework** like Payload vs. Medusa:

> "Schema in code" like strapi, payload, sanity and others. No need to manually add api endpoints, types, validity checks, UI. data import and export.

This is key to offer the most **easily** customizable commerce framework. Medusa already has incredible commerce modules and a "full-stack" commerce platform (backend+admin+storefront). But one could argue that [launching](https://payloadcms.com/blog/launch-week-day-5-ecommerce-starter-kit) and customizing an [e-commerce](https://payloadcms.com/blog/how-to-build-an-e-commerce-site-with-nextjs) platform on Payload is [easier](https://payloadcms.com/developers).

Taking this even further, one can envision a future where non-developers use the framework to build custom entities, create workflows and can do most things via the Admin UI. A good example of this is Directus's [Data Studio](https://docs.directus.io/app/data-model.html) and [Flows](https://docs.directus.io/app/flows.html) (very similar to Medusa Workflows).

Super excited about the direction of Medusa and the upcoming 2.0 release! 🪼🚀

**Comment by lakshmajee** [comment]
It is great effort to put in this project. I appreciate it.

1. Module / package decoupling is really a needed feature. 
2. Package re-organization is not necessarily needed IMO, fancy package naming can be done later after a year. No problem with that.
3. Leaving previous stable release (1.x) with out any well maintained version (2.x-rc) will create a bad impression on the project trustiness with in community or developers. (Im already feeling it after started using it for almost 7 days).
     3.1. Leaving open issues on v1.x and assigning low priority makes it worse.
     3.2. Marking low priority on community developers PR's is really a bad. it discourages them.
     
Thank you 😊

**Reply by luluhoc** [reply]
@awkward-minion Now the PR's getting closed for 1.x, I rewrote the SendGrid plugin to typescript, and it was just closed. It could have been merged to 1.x tho...

**Comment by NicolasGorga** [comment]
@olivermrbl First of all, thanks for the great work.

Will there be more customization possibilities with 2.0?

Things I think are necessary in order to use the shipped Admin UI in a multi vendor marketplace env:

- Show hide left vertical menu items based on a user role (or any other logic that is relevant to the business). Example: Vendors should be able to see a subset of the items that an Admin of the system (like me), which should be able to see and configure everything related to the system and the individual Vendors

- Add elements to the left vertical menu: Correct me if i'm wrong (I've only been using Medusa for less than a week), but i think there is no way to add items to the left hand side vertical menu, to route to custom pages. Example: as an Admin, i would like to add a "Vendors" item, that navigates me to a page i would create to see the registered Vendors, to approve them or dissaprove them, as well as configure things like sale_fee, as for one Vendor i woul maybe want to configure a 10% fee to every purchase, will another might have a 15% fee.

- Customize the route that should be navigated after login: right now it is fixed to the dashboard. Example: If a Vendor tries to log in, but hasn't been approved by me yet, they should be navigated to a '/pending-verification' route. Tied to this, the point below.

- Ability to define guards for routes shipped with Admin UI: Example: If a Vendor that hasn't been approved by me yet tries to navigate to any route, i would like them to be redirected automaticall to a route like '/pending-verification'

If you need any more input regarding the limitations I have identified for Marketplace you have my full attention.

Greetings,
Nicolas

**Reply by danvayn** [reply]
You should share these in discord.

@olivermrbl did you see this reply

**Comment by HarleySalas** [comment]
Perhaps it's already possible, but I'd love to see some features able to be *disabled* in the admin ui. 

For example, we often have clients with simpler requirements. Perhaps they don't need something like pricing overrides, or gift cards. I don't like shipping anything unless every feature is usable. I hate the idea of "heres your dashboard, just don't touch that button," or "that doesn't work properly." Even Regions are something that many of our clients don't need. It'd be nice to just throw a default region in and delete the concept from inside the dashboard. 

Perhaps there is already an easy way to do it, I'm just unaware.

**Reply by DexA1948** [reply]
open node modules and hide from there, why would you suggest removing gift cards tho?

**Reply by HarleySalas** [reply]
It's just an example. There are cases where clients don't want that feature, or it doesn't apply to the scope of the product, or store's concept. So, why should there be something in the UI, which isn't implemented, or used whatsoever?

Simply hiding it via node_modules is definitely not a good pattern. If that sort of feature never makes it, so be it. I was just providing some feedback.

**Comment by visualies** [comment]
Im checking the website daily at this point, cant wait for 2.0 🤞

**Reply by poudelprakash** [reply]
same here! at lease pre-release versions should have been available for developers to get into

**Comment by artur1214** [comment]
It's July already, can't wait for RC )

**Reply by Bollznet** [reply]
There is news??

**Reply by SalahAdDin** [reply]
Not yet.

**Comment by amaster507** [comment]
Please see [PayPal Payment Gateway Package for Medusa v2](https://github.com/amaster507/medusa-payment-paypal) that is ready for testing

https://github.com/medusajs/medusa/discussions/9346

**Comment by SalahAdDin** [comment]
@olivermrbl Is there any guide to migrate plugins?

**Reply by Megrax** [reply]
For file-storage-related plugins, here is the migration solution: switch to using the **S3 File Module Provider**.
https://docs.medusajs.com/resources/architectural-modules/file/s3

**Reply by SalahAdDin** [reply]
> For file-storage-related plugins, here is the migration solution: switch to using the **S3 File Module Provider**. https://docs.medusajs.com/resources/architectural-modules/file/s3

Is it compatible with Cloudinary?

**Comment by hariDasu** [comment]
i am interested in using supabase-storage plugin and supabase as the database for medusa... is this still possible with 2.0 arch?

**Reply by aaa3334** [reply]
Have set up the database and storage with 2.0 - its just a bit of a different setup (basically just the env variables and medusa.config). I haven't used v1 though

**Reply by joaoemersonufc** [reply]
How do you make the setup @aaa3334? I encounter an error when I do the npm install

```
npm error Found: @medusajs/medusa@2.3.1
npm error node_modules/@medusajs/medusa
npm error   @medusajs/medusa@"^2.3.0" from the root project    
npm error
npm error Could not resolve dependency:
npm error peer @medusajs/medusa@"^1.20.2" from medusa-storage-supabase@0.1.7
```

**Reply by aaa3334** [reply]
I set it up with the url key to the database. For storage I followed this: https://docs.medusajs.com/resources/architectural-modules/file/s3#register-the-s3-file-module

I think the medusa-storage-supabase is for the old medusa version (you can see it requires version 1, not 2 in your error code)
